config/consensus.go

```diff
@@ -387,6 +387,11 @@ type ConsensusParams struct {
 	// StateProofStrengthTarget represents either k+q (for pre-quantum security) or k+2q (for post-quantum security)
 	StateProofStrengthTarget uint64
 
+	// StateProofRecoveryInterval represents the number of state proof intervals that the network will try to catch-up with.
+	// When the difference between the latest state proof and the current round will be greater than value, Nodes will
+	// release resources allocated for creating state proofs.
+	StateProofRecoveryInterval uint64
```

Comment:
I think this should be a consensus param since different values would on different node might cause nodes to disagree on stateproof transactions (they might remove data used for the verification)

Reply:
@gmalouf @cce, We would like to hear your thoughts on this.

Reply:
Here is why I think this parameter should be a configuration parameter, and not a consensus parameter:

This parameter decides how much memory a system can tolerate. This parameter needs to be set to a reasonably large value, to provide some tolerance in case the SP is delayed, and perhaps give enough time for intervention when something wrong happens.

However, a machine with limited amount of memory will crash when the memory usage goes up because of this parameter.

As a more concrete example, let's say there are two machines, one with 95% stake and a lot of memory, and another machine, with 5% stake, and limited memory. It is not a good idea to have the same value for this parameter to all of these machines. 

The machines with high stake and high memory, can and should set a large value for this parameter, while others, with limited memory, can only tolerate a small value. 

Therefore, I think this parameter should not be a consensus parameter. 

@gmalouf @brianolson your thoughts.

Reply:
How big is a state proof? (serialized? in memory?)
I'm not sure I understand the work flow, but if something asks `algod` to validate against state proofs, does it return three answers {valid, invalid, I don't know} ? If "I don't know" is an acceptable answer, then better nodes hold more, cheaper nodes hold fewer.

Reply:
a key observation in our discussion is that there are two parts of data needed for stateproof
1 - data needed to create a stateproof transaction (participation array, bound of signatures etc)
2- data needed to verify a stateproof transaction on-chain. (need to block header from 256 rounds before the stateproof)

and currently, the new parameter affects both parts 
In my opinion, regarding the second part - all node must be synced on what blocks they have in order to verify a stateproof transaction. Having different parameter values might lead to disagreements.

However, regarding the first part, we can definitely extend this and allow some beefy servers to store more data (this will be implemented using a configuration parameter and not consensus ). This is part of what we plan to implement next quarter.

really hope this addresses your concern

Reply:
Is it really only relay nodes that have to worry about this? If so, we should ensure their minimum recommended specs can hold up over some delayed period of reason.

Reply:
No. All nodes that have online accounts. This parameter will impact their ability to sign blocks with SP transaction in them.

Reply:
Yes, we need a live discussion for this, absolutely. And the value of 10, which is too much for Brian :-), is of very limited use if set that low. 
Also, once added a consensus parameter, it will be there forever!

Reply:
@algoidan says:
> > a key observation in our discussion is that there are two parts of data needed for stateproof
> > 1 - data needed to create a stateproof transaction (participation array, bound of signatures etc)
> > 2- data needed to verify a stateproof transaction on-chain. (need to block header from 256 rounds before the stateproof)
> >
> > and currently, the new parameter affects both parts
> > In my opinion, regarding the second part - all nodes must be synced on what blocks they have in order to verify a stateproof transaction. Having different parameter values might lead to disagreements.

@algonautshant replies:
> This is the argument for having it as a consensus parameter. But I would say that, accounts will either vote yes, or no to the transaction/block. If enough "yes" votes are obtained, then all is good. The idea that all nodes must be "synced", is not going to happen.
>
> In the event of a delayed SP, either all will say "No" (parameter too small), or the weak ones will crash, and others will say "Yes" (parameter big enough).
>
> No. All nodes that have online accounts. This parameter will impact their ability to sign blocks with SP transaction in them.

I agree with Idan on #2. There cannot be disagreement over whether a transaction is valid. All nodes (whether relay, participating, non-participating) must be able to validate transactions and block proposals. I disagree with "If enough 'yes' votes are obtained, then all is good. The idea that all nodes must be 'synced', is not going to happen."

My understanding is that consensus means every node can validate all transactions, proposals, votes, and blocks for themselves. Relaxing this requirement introduces subjectivity into whether block proposals or transactions are valid, and would lead to consensus failing to find agreement (and stalling) if enough nodes disagree on the validity of the same transactions/blocks.

@algoidan goes on:
> However, regarding the first part, we can definitely extend this and allow some beefy servers to store more data (this will be implemented using a configuration parameter and not consensus ). This is part of what we plan to implement next quarter.

Would it help to give this parameter a different name perhaps?

Reply:
Talking this over with Shant just now, `StateProofInterval` and `StateProofRecoveryInterval` _must_ be consensus parameters because they govern evaluating new transactions in a block (the recording of a state proof is a new txn type) and everyone must have the same rules for evaluating that txn. *Also* this commits us to some extra storage on each node (at least (32 byte addr)*(1000 top accounts)*(10 state proof periods) == 320,000 bytes) to store which accounts were valid signers for each state proof we may yet need to check signatures for. I was going to say 'and we should have a discussion about whether to commit that storage', but 320kB ought to be fine.